src/preact/component/3p-frame.js

```diff
@@ -0,0 +1,139 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {IframeEmbed} from './iframe';
+import {dict} from '../../utils/object';
+import {forwardRef} from '../compat';
+import {
+  generateSentinel,
+  getBootstrapUrl,
+  getDefaultBootstrapBaseUrl,
+} from '../../3p-frame';
+import {
+  getOptionalSandboxFlags,
+  getRequiredSandboxFlags,
+} from '../../core/3p-frame';
+import {parseUrlDeprecated} from '../../url';
+import {sequentialIdGenerator} from '../../utils/id-generator';
+import {useLayoutEffect, useRef, useState} from '../../../src/preact';
+
+/** @type {!Object<string,function>} 3p frames for that type. */
+export const countGenerators = {};
+
+/** @enum {string} */
+export const MessageType = {
+  // TODO(wg-bento): Add more types as they become needed.
+  EMBED_SIZE: 'embed-size',
+};
+
+// Block synchronous XHR in ad. These are very rare, but super bad for UX
+// as they block the UI thread for the arbitrary amount of time until the
+// request completes.
+const BLOCK_SYNC_XHR = "sync-xhr 'none';";
+
+// TODO(wg-bento): UA check for required flags without iframe element
+const DEFAULT_SANDBOX =
+  getRequiredSandboxFlags().join(' ') +
+  ' ' +
+  getOptionalSandboxFlags().join(' ');
+
+/**
+ * Creates the iframe for the embed. Applies correct size and passes the embed
+ * attributes to the frame via JSON inside the fragment.
+ * @param {!IframeEmbedDef.Props} props
+ * @param {{current: (!IframeEmbedDef.Api|null)}} ref
+ * @return {PreactDef.Renderable}
+ */
+function ProxyIframeEmbedWithRef(
+  {
+    name: nameProp,
+    options = {},
+    messageHandler,
+    sandbox = DEFAULT_SANDBOX,
+    src: srcProp,
+    type,
+    title = type,
+    ...rest
+  },
+  ref
+) {
+  const contentRef = useRef(null);
+  if (!countGenerators[type]) {
+    countGenerators[type] = sequentialIdGenerator();
+  }
+  const [count] = useState(countGenerators[type]);
+  const [name, setName] = useState(nameProp);
+  const src = useRef(null);
+  const win = contentRef.current?.ownerDocument?.defaultView;
```

Comment:
This is problematic because 
(a) It won't be available here on the first render.
(b) It won't be available in SSR mode at all.

What I think we should do is use the most default version of the bootstrap URL. E.g. resolved on AMP CDN or whichever we do. And then we can leave it to AMP Layer to provide it using `win` via `srcProp`. And then we can, over time, improve the configuration of this prop. For instance, in the future you may decide that the best option is to use some `ProxyIframeConfigContext/Provider`.

Reply:
Since `win` is needed for a number of properties here, including parts of the iframe name, could you recommend some sensible defaults for these? Especially since the default case for `getDefaultBootstrapBaseUrl` relies on `internalRuntimeVersion`, I'm not sure a good way to hardcode a default that doesn't need to be rotated?

Reply:
That does sound like a good idea for a `ProxyIframeEmbed`.